org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/format/LSPFormatter.java

```diff
@@ -25,94 +23,77 @@
+import org.eclipse.lsp4e.LanguageServers.VersionedEdits;
 import org.eclipse.lsp4j.DocumentFormattingParams;
 import org.eclipse.lsp4j.DocumentRangeFormattingParams;
 import org.eclipse.lsp4j.FormattingOptions;
 import org.eclipse.lsp4j.Position;
 import org.eclipse.lsp4j.Range;
 import org.eclipse.lsp4j.ServerCapabilities;
-import org.eclipse.lsp4j.TextEdit;
+import org.eclipse.lsp4j.TextDocumentIdentifier;
 import org.eclipse.ui.editors.text.EditorsUI;
 import org.eclipse.ui.texteditor.AbstractDecoratedTextEditorPreferenceConstants;
 
 public class LSPFormatter {
-
-	public record VersionedFormatRequest(int version, CompletableFuture<List<? extends TextEdit>> edits) {
-		public VersionedFormatRequest() {
-			this(-1, CompletableFuture.completedFuture(Collections.emptyList()));
+	public CompletableFuture<Optional<VersionedEdits>> requestFormatting(@NonNull IDocument document, @NonNull ITextSelection textSelection) throws BadLocationException {
+		URI uri = LSPEclipseUtils.toUri(document);
+		if (uri == null) {
+			return CompletableFuture.completedFuture(Optional.empty());
 		}
-	}
+		LanguageServerDocumentExecutor executor = LanguageServers.forDocument(document).withFilter(LSPFormatter::supportsFormatting);
+		FormattingOptions formatOptions = getFormatOptions();
+		TextDocumentIdentifier docId = new TextDocumentIdentifier(uri.toString());
 
-	private @NonNull LSPDocumentInfo getLSPDocumentInfo(Collection<@NonNull LSPDocumentInfo> infos) {
-		// TODO consider a better strategy for that, maybe iterate on all LS until one
-		// gives a result
-		return infos.iterator().next();
-	}
+		DocumentRangeFormattingParams rangeParams = getRangeFormattingParams(document, textSelection, formatOptions,
+				docId);
+
+		DocumentFormattingParams params = getFullFormatParams(formatOptions, docId);
 
-	public void applyEdits(@NonNull IDocument document, List<? extends TextEdit> edits, int version)
-			throws ConcurrentModificationException {
-		Collection<@NonNull LSPDocumentInfo> infos = LanguageServiceAccessor.getLSPDocumentInfosFor(document,
-				LSPFormatter::supportsFormatting);
-		if (getLSPDocumentInfo(infos).getVersion() != version) {
-			throw new ConcurrentModificationException();
-		} else {
-			try {
-				LSPEclipseUtils.applyEdits(document, edits);
-			} catch (BadLocationException e) {
-				LanguageServerPlugin.logError(e);
+		// TODO: Could refine this algorithm: at present this grabs the first non-null response but the most functional
+		// implementation (if a text selection is present) would try all the servers in turn to see if they supported
+		// range formatting, falling back to a full format if unavailable
+		return executor.computeFirst((w, ls) -> {
+			final ServerCapabilities capabilities = w.getServerCapabilities();
+			if (isDocumentRangeFormattingSupported(capabilities)
+					&& (!isDocumentFormattingSupported(capabilities)
+							|| textSelection.getLength() != 0)) {
```

Comment:
I think it would be slightly easier to understand if we write && !(textSelection.getLength() == 0 && isDocumentFormattingSupported(capabilities))